### ISSUE TITLE
Numeric Tower / Union[float, int]

### ISSUE THREAD

**Issue by KyleKing** [beartype/beartype#66]
I think I found an edge case with the annotation `float` when passed an `int` in `beartype`. For convenience, PEP484 accepts int when `float` is annotated: https://www.python.org/dev/peps/pep-0484/#the-numeric-tower

```py
from beartype import beartype


@beartype
def measure_cave(length: float, width: float):
    return length * width


measure_cave(12.0, 12.0)  # PASS
measure_cave(12, 12)  # type hint <class 'float'>, as "12" not instance of float.
```

I'm currently using `Union[float, int]` to work with beartype and wondered if you would consider implementing the PEP484 "numeric tower" logic or keep beartype as-is?

**Comment by leycec** [comment]
**This is such a great issue.** I'm so grateful to our crack commando squad of black-ops typing aficionados for repeatedly weighing in. They say what I cannot, for they are smart and I am tired. This is what happens when your eyes get encrusted with the detritus of sleep deprivation, folks: you wait for @posita and @TeamSpen210 to pierce the darkness for you.

## tl;dr

Yup, we're prolly sticking with either:

``` python
# For those who worship before the graven alter of the official numeric tower:
from numbers import Number

# For those who desecrate idolatrous effigies of false abstract base classes:
from typing import Union
Number = Union[int, float, complex]
```

I am so, so sorry for what I have done to your codebase.

## Why Are You So Horrible, @leycec?

Yes, @leycec *is* a **B**ad **E**nough **D**ude (BED).

![that's me](https://i.kym-cdn.com/entries/icons/original/000/001/619/Are_You_A_Bad_Enough_Dude_banner.png)

With extreme apologies to AI King @KyleKing, <sup>*please don't send the robot dogs*</sup> PEP 484, mypy, and (yes) Guido himself, `@beartype` would prefer to selectively pretend that something that was standardized wasn't standardized. That is to say, we're pretty sure Guido and those who approved PEP 484 got it wrong there.

As @posita himself unearthed from Guido's inbox, <sup>*DropBox power!*</sup> Guido himself now acknowledges the ubiquitous world-straddling girth of [PEP 3141](https://www.python.org/dev/peps/pep-3141). PEP 3141 came first and PEP 484 failed to deprecate PEP 3141; ergo, PEP 3141 takes precedence. Suck it, PEP 484.

## Non-Euclidean Numeric Tower, We Invoke Thee!

**So.** Here we stand on the crumbling shores of a long-forgotten derelict civilization. The principal issues with PEP 3141 and the standard `numbers` module are (in increasing order of impotence):

* **Everyone ignored it,** because no one knew it even existed.
* **Stack type checkers still ignore it,** because they are dumb. <sup>score 1: `beartype`</sup>
* **Protocols invite hidden `O(n)` costs,** which is admittedly horrifying. Back before I became disillusioned by the ~~veil of Maya~~ gritty facade of reality, I was once pretty sure that `ABCMeta` efficiently cached (and thus effectively amortized) those costs. Having just reexamined the `typing.py` implementation in Python 3.10, I no longer know anything. Those costs might very well *not* be cached, in which case we can only collectively sigh and fling ourselves into the sea.

**So.** `numbers` had an advertising problem. But that wasn't the fault of `numbers`; that was the fault of the humans surrounding `numbers`. Humanity, you have much to answer for.

**So.** `numbers` probably also has an efficiency problem. That absolutely *is* the fault of `numbers` and (more generally) the standard `abc` module underlying `numbers`. Remind me to never use `ABCMeta` in `beartype`. <sub>←oh gods we already did it 5 times</sub>

## What Could Possibly Go Wrong? Everything. Everything Could.

The principal issue with transparently treating `float` as `int` *is that it's wrong* – especially in the numeric contexts that `@beartype` increasingly finds itself deployed to (e.g., finance, machine learning).

Floats aren't integers. Integers aren't floats. Specifically, there exist a countably infinite number of integers with *no* precise floating-point representation; likewise, there exist an *un*countably infinite number of floats with *no* corresponding integer representation.

Floats and integers are *not* losslessly coercible is what I'm sayin'. Converting one into the other results in a loss of precision in both the average and worst case. This is why numeric frameworks like NumPy have recently moved from implicit to explicit `float` <-> `int` coercions. Previously, NumPy implicitly coerced dtypes to and from `float` and `int`; now, NumPy requires you to explicitly permit NumPy to do that if you want NumPy to do that.

This is a good thing. [Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020), because we are all enlightened light beings here. You may now be thinking: "Go to Heck, @leycec! You go to Heck and sleep fitfully there."

**Okay.** Stay with me, folks! You can punch me in my fat gullet in five minutes if you remain unconvinced. <sup>would i lie</sup>

If `beartype` implicitly accepts integers for all `float` type hints, then **`beartype` prosumers will no longer be able to prevent their userbase from erroneously passing integers.** Right? In 2021, most industrial-strength APIs *really* want to prevent their userbase from erroneously passing integers where floats are required. Currently, `beartype` supports this sensible constraint out-of-the-box at runtime – which is *far* more valuable than a static type checker supporting this sensible constraint out-of-the-box at static type-checking time.

If `beartype` implicitly accepts integers for all `float` type hints, then `beartype` prosumers will have to resort to beartype validators to force `beartype` to revert back to its prior sensible behaviour: e.g.,

```
from beartype.vale import Is
from typing import Annotated

# Type hint matching any non-integer float. Raise your voice in anger, userbase!
FloatNonint = Annotated[float, Is[
    lambda number: not isinstance(number, int)]]
```

## Don't Believe Me. Believe Those Who Use Me.

[`alibi`, a popular ML-splaining framework (*it's like mansplaining, only done by AI*), wants to use `beartype`](https://github.com/SeldonIO/alibi/pull/511). This use is contigent on `beartype` behaving sanely.

Currently, `beartype` behaves sanely by raising human-readable exceptions when `@beartype`-decorated callables annotated as accepting floats are passed integers. There are, like, a *hundred* of these test-time unit test failures, and stuff:

```
beartype.roar.BeartypeCallHintPepParamException: @beartyped CounterfactualProto.__init__() parameter gamma="100" violates type hint <class 'float'>, as "100" not instance of float.
```

Curiosity maimed the honey-addicted bear, so [I noted that and suggested they might globally replace `float` with `numbers.Real` in their type hints](https://github.com/SeldonIO/alibi/pull/511#issuecomment-951668473). Turns out [they meant what they said and said what they meant. They actually *did* meant `float`, because all of their callables only accept floats; they do *not* accept integers](https://github.com/SeldonIO/alibi/pull/511#issuecomment-951719674):

> Nice catch! I think this is an example where the parameter really should've been a `float`, `beartype` paying dividends already.

Are you gonna break `alibi`? I ain't gonna break `alibi`. I like my kneecaps where they are – loosely attached to my legs.

## Cost-Benefit Analysis

Now we're into the hard stuff. Dust off those Grandma-era liquor cabinets, because we are going there.

### These Are the Costs of an Irresponsible Life

Costs of `@beartype` *not* implicitly accepting integers for `float` type hints:

* **RSI.** Users type an additional 12 characters. Justifiable lawsuits ensue. `@beartype` collapses under a rising tide of unpaid legal fees and collective Internet outrage.

And... that's about it. "But, wait! You calloused fool! What about mypy? Won't someone think about the static type-checker?", you are now venting. I can hear you over there, by the way.

Very well, **we don't do what static type checkers do.** But that's permissible here, because static type checkers are (arguably) *too* permissive here. Mypy won't emit errors or warnings for callables type-hinted with `Union[float, int]` instead of `float`, because the former is semantically equivalent to the latter from the static type-checking perspective.

So. The only cost remains RSI. You've got it. I've got it. We've all got it. What's 12 characters more when your wrist has already been ground down into a thin bone spur of unadulterated pain?

### These Are the Benefits of an Irresponsible Life

Benefits of `@beartype` *not* implicitly accepting integers for `float` type hints:

* **Users catch nasty bugs.** Does that sound like the principal benefit of type-checking to anyone else? Is it just me? Yes, we dared to utter the unthinkable! Passing an `int` where a `float` was expected *is* a bug in the general case, because: <sup>*cue that acoustic guitar campfire sing-a-long*</sup>
  * *Integers are not floats.*
  * *Floats are not integers.*

## These Are the Conclusions of the Starship Bearaprise

On the one hand, we're giving everyone RSI. On the other hand, we're doing our jobs.

## And... We're Goin' Back to PEP 484

Anyone else notice [the PEP 484 subsection we're debating is *super*-weird?](https://www.python.org/dev/peps/pep-0484/#toc-entry-18) First, there's this choice nugget:

> There are some issues with these ABCs [i.e., `numbers`]...

There aren't, actually. Okay, *okay*. `O(n)` runtime complexity *is* a problem – but not for most techbros. Anyone else notice how none of the supposed issues were actually enumerated or even mentioned? Guido throwin' some baseless shade right there. Then, there's this even choicer nugget:

> ...but the built-in concrete numeric classes `complex`, `float` and `int` are ubiquitous (especially the latter two :-).

**OMG.** There is *not* an ASCII emoji in my favourite PEP standard. There just *isn't.* But... there is. Anyone else notice how this entire subsection reads more like casual email-speak than the quasi-format standards-speak of the rest of the PEP? It's almost like... no one actually even peer-reviewed or edited this subsection. *Just sayin'.*

We continue on our feeble death march with this frank admission:

> Rather than requiring that users write import `numbers` and then use `numbers.Float`, etc...

Yes? And? Why *aren't* we doing that? Because that's what we should be doing.

> ...this PEP proposes a straightforward shortcut that is almost as effective:

**Oh, bollocks.** PEP subsection straight-up admitted it adopted a suboptimal, less effective solution than the previously standardized solution – *just 'cause.* No demonstrable justifications are given. Guido himself now acknowledges that `numbers` is preferable.

**So, bollocks.** We defy the established order, because we are `beartype`. We are preserving a distinction between `float` and `int`, because there *is* a distinction between `float` and `int`. The two are *not* transparently interchangeable and any attempt to treat them as such invites type safety violations and the immanent destruction of the supposedly incomplete Death Star. <sup>←brought to you by `beartype`</sup>

## A Pox on Beartype's House

That said... *let's leave this open for all eternity.*

There are extremely valid insights on bold sides of the opinion-littered back alley here. I invite them! Come! Berate me, excoriate my avatar, burn a miniature effigy of [Mr. Nectar Palm (the @beartype mascot, yo)](https://github.com/beartype/beartype-assets/tree/main/banner) on your front lawn and then tweet us an animated GIF of the ensuing wreckage!

Do this in rembrance of this issue. Someday, even @leycec too may change his unsubstantiated worldview on QA.

**Comment by KyleKing** [comment]
Thanks @leycec, @posita, and @TeamSpen210! I feel much more enlightened now and I think we can safely lock the numeric tower and pretend it never existed 😊

**Comment by posita** [comment]
_**Yasss!**_ Casual ~erudite advice~ _streams of consciousness_ from influencers masquerading as standards that disproportionately muddle the lives of hard workin' developers everywhere can _[suck it](https://github.com/beartype/beartype/issues/65#issuecomment-955712840)_! _¡Viva la osolución¹!_

[![Yasss!](https://www.utilitygift.co.uk/media/catalog/product/cache/f3737f4db7caf1a8b5e4e98cdcf0ed8c/y/a/yaaas.jpg)](https://www.utilitygift.co.uk/yasss-61581)

@KyleKing, I'm still working on something that will help you out on the Mypy front. I might need another day or two, but I'll post here once it's up.

---

¹ _Osolución_ was intended as a portmanteau of _oso_ (meaning bear) and _revolución_ (meaning revolution). Native Spanish speakers are going to have to help me out on this one. I don't even know whether Spanish has portmanteaus of this kind. I can't seem to find a Spanish word that describes one. _Maleta de palabras_ (word suitcase), perhaps?

**Comment by posita** [comment]
All right, I've now made available a [_**super**_ early VIP sneak preview of ``numerary``](https://posita.github.io/numerary/)! Which probably doesn't solve a problem you have, @KyleKing, but I used your issue as an excuse to tinker with it instead of doing real work.

The documentation is definitely on the rough side but you _**should**_ be able to do something like:

``` python
from typing import TypeVar
from beartype import beartype
from numerary import RealLike

_RealT = TypeVar("_RealT", bound=RealLike)

@beartype
def measure_cave(length: _RealT, width: _RealT) -> _RealT:
    return length * width

assert measure_cave(12, 12) == 144
assert measure_cave(12.0, 12.0) == 144.0

from fractions import Fraction
assert measure_cave(Fraction(12), Fraction(144, 12)) == Fraction(144)

from decimal import Decimal
assert measure_cave(Decimal("12.0"), Decimal(12)) == Decimal(144)

try:
    # Mypy will properly detect these mismatched arguments
    measure_cave(Fraction(12), Decimal(12))
except TypeError:
    assert True, "Decimals don't play well with others"
else:
    assert False, "expected exception was not thrown"

# Mypy (and beartype) will treat this as okay (your original issue)
assert measure_cave(12.0, 12) == 144.0
```

You can also do things like:

``` python
from typing import TypeVar, Union
from beartype import beartype, roar
from numerary import IntegralLike

_IntegralT = TypeVar("_IntegralT", bound=IntegralLike)

@beartype
def shift_them_bits(
    arg: _IntegralT, amount: Union[_IntegralT, int]
) -> _IntegralT:
    if int(amount) < 0:  # IntegralLike includes SupportsInt
        return arg >> abs(amount)
    else:
        return arg << amount

assert shift_them_bits(1, 1) == 2

try:
    assert shift_them_bits(2.0, 1) == 4  # Mypy error
except roar.BeartypeException:
    assert True, "nice catch!"
else:
    assert False, "expected exception was not thrown"

try:
    from sympy import sympify
    assert shift_them_bits(sympify("2"), sympify("2")) == sympify("8")
except ImportError:
    pass
```

**Comment by leycec** [comment]
Bravo! Bravura! Bravissimo! And other sketchy loan words that probably don't mean what we think they mean! First, this:

> `dyce` - a pure-Python library for modeling arbitrarily complex dice mechanics and **~~mother~~ _birthing code base_** 🙄 of `numerary`!

My calcified pineal gland now needs a deep cleanse. On reading the ominous occultic jargon "birthing code base," is it common for an autumnal chill to crawl in slo-mo across my sweat-slicked skin?

That's probably common. Next, this:

```
In [13]: %timeit -r10 isinstance(val, Rational)
326 ns ± 1.26 ns per loop (mean ± std. dev. of 10 runs, 1000000 loops each)
```

...*seems* to be missing a preceding `val = ` assignment? I dunno. I'm just the unkempt bear feeder.

> That’s *forty times* slower. 😶 And that’s just with a two-property protocol. How much worse would it be if we had enumerated all the dunder methods? 😰

`</gulp>`

I'm on the edge of my seat here, bro. I don't know and I didn't know I wanted to know, but now I know that I really did want to know all this time. How much worse *would* it have been? I'm guessing, "Lots. I've seen crippling performance concerns the likes of which you people wouldn't believe."

Next, this:

``` python
SupportsRealComparisonsNotComplexLies = Annotated[
  SupportsRealComparisons, Is[lambda arg: not isinstance(arg, complex)]
]
```

![oh. shit.](https://www.hellokpop.com/wp-content/uploads/2019/12/mic-drop-600m-430x400.jpg)

**Comment by posita** [comment]
Sage advice beyond my wildest dreams. Please accept my many thanks and know I am in the process of following it (for some definition of "following"). Selective responses below….

> ..._seems_ to be missing a preceding `val = ` assignment? I dunno. I'm just the unkempt bear feeder.

Solid catch. Thank you!

> I'm on the edge of my seat here, bro. I don't know and I didn't know I wanted to know, but now I know that I really did want to know all this time. How much worse _would_ it have been? I'm guessing, "Lots. I've seen crippling performance concerns the likes of which you people wouldn't believe."
> …
> Actually, it'd be even beautifuller if you could casually throw out some `timeit` timings here contrasting the abyssal performance of the standard `typing.Protocol`-based approach against all _three_ of the `numerary` approaches (i.e., your optimized `typing.Protocol`-based API, your short-circuiting `typing.Union`-based API, and your short-circuiting tuple-based API). It would be ideologically fascinating and pragmatically exemplary if your optimized `typing.Protocol`-based API could get within an order of magnitude of the two short-circuiting alternatives.

I can't claim to have witnessed any attack ships near the shoulder of Orion (on fire or not) nor have I seen glittering c-beams (wherever they may be found). But it _was_ my intention to expand upon this and other (currently unhelpfully nebulous) performance commentary. Stay tuned … .

> Pretty sure I'm the geriatric white-haired guy in the passenger seat while youthful svelte @posita up on the roof chuckin' dynamite around with emoji mittens.

My family might peg me closer to Murdoch, but maybe that's just my affinity for trash bags. Not sure.

> … Yet no one thought of [caching] – until now.

I'm pretty sure that's not true. Consider:

> * Back before I became disillusioned by the ~veil of Maya~ gritty facade of reality, I was once pretty sure that `ABCMeta` efficiently cached (and thus effectively amortized) those costs. Having just reexamined the `typing.py` implementation in Python 3.10, I no longer know anything. Those costs might very well _not_ be cached, in which case we can only collectively sigh and fling ourselves into the sea.

Now I can't quite remember where I heard that or who said it, so I'm quoting loosely from memory (including markup). I'm pretty confident I'm not the only one with that idea … . 😉

> Next, I couldn't help myself. Because caching is joy, I delved into [the joy you implanted into your `numerary.types` submodule](https://github.com/posita/numerary/blob/3abc6f8fde96f6260b13a6e9b1155e39c3e3c204/numerary/types.py#L106). … Although technically fine, `self` should probably be `cls` here for clarity.

Yeah, I have no idea what I was thinking there.

> … Additionally, the `cls._abc_inst_check_cache[cls]` seems redundant there. I preserved it because kinda unclear as to why the double-check exists. Protocol subclasses, maybe?

Actually, you're right to question that. But that makes both of us wrong (or at least needlessly inefficient). This is where it ended up after your observation:

``` python
class CachingProtocolMeta(_ProtocolMeta):
  # …
  def __new__(…) -> _TT:
    cls = super().__new__(mcls, name, bases, namespace, **kw)  # type: ignore
    cache: Dict[Tuple[type], bool] = {}
    cls._abc_inst_check_cache = cache
    return cls
  def __instancecheck__(cls, inst: Any) -> bool:
    inst_t = type(inst)
    if inst_t not in cls._abc_inst_check_cache:
      cls._abc_inst_check_cache[inst_t] = super().__instancecheck__(inst)
    return cls._abc_inst_check_cache[inst_t]
```

Again, I don't know what I was thinking. In all likelihood, I wasn't … at least not enough.

> ...is no longer strictly true. Praise the Sun! Thanks entirely to [PEP 604 -- Allow writing union types as X | Y](https://www.python.org/dev/peps/pep-0604) and the recently released Python 3.10 standardizing PEP 604, both of the following are now permissible under Python ≥ 3.10 – which admittedly doesn't matter for most devs yet, but _should_ matter in two or three years, when Python 3.10 becomes the default for most Linux distros: …
>
> If you (like me) just blew a non-redundant O-ring, allow me to supply you with this only mildly used sanitary towelette from the back.

I too became super excited for PEP 604 goodness because it approaches something sane and unified. Then I got even exciteder when I read [this](https://github.com/python/mypy/pull/9647#issue-729964835):

> This PR attempts to add the original implementation for [PEP 604](https://www.python.org/dev/peps/pep-0604/) … . The PEP will be added to Python in `3.10`. [S]ince it's already acceptable syntax, we can even use it for older versions with `from __future__ import annotations`, in type comments or stub files. [python/typing#429 (comment)](https://github.com/python/typing/issues/429#issuecomment-690594334)
> 
> **Example**
> 
> ```python
> # python 3.7
> from __future__ import annotations
> from typing import Union
> 
> t1: Union[int, str]
> t2: int | str
> ```

But then I read [this](https://github.com/python/mypy/issues/9610#issuecomment-711147567):

> What doesn't work are type aliases. The following would cause a type error:
> 
> ```python
> from __future__ import annotations
> 
> T1 = str | None  # TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
> d: T1 = "Hello World"
> ```

![Sad T pities the fool](https://c.tenor.com/QE3uXKXMA74AAAAM/cry-tears.gif)

> … If I'm awesomer than `beartype`, that means `beartype` is strictly less than something else and thus provably deficient. Why, `beartype`? Why haven't you yet surpassed your progenitors and then quietly silenced them on an otherwise tranquil summer evening that will be known only as The Night of the Long Claws? 🐾

Okay, hear me out. @leycec _made_ ``beartype``, and not in a donated-some-genetic-material-and-nature-mostly-handled-the-rest kind of way, but more of a dr-eldon-tyrell way … . Wait. Oh shoot. Maybe that's what you were getting at and we're not quite to that part of the story … . Let's hope it takes a different trajectory.

But you are _also_ making ``numerary`` better! Wait. Shoot. So is ``beartype``. Man that _is_ a close race for awesomest. I (perhaps stubbornly) stand by my original statement.

> … Allow me to snip out the relevant words into a vast tapestry of woe:
> 
> > `beartype` ... don't want... to use it... for performance reasons
> 
> _this. this cannot be._

Okay, okay, how about "… don't want ... to use it ... for ~~performance~~ _some strange_ reason[] …"? 😜

> … "Please, Matt. Straight-up switch the GLWTS for [the same license that you've released `dyce` under. So, **MIT** is what I'm saying.](https://github.com/posita/dyce/blob/master/LICENSE)"

This is solid advice. In retrospect, GLWTS was definitely not the best choice on my part. Can you tell I was feeling a little punchy? I'll bet you could. I'll bet that came through. I'll even wager it would have _still_ come through had I omitted all the symbolic swearing. You have helped me engage in the helpful strategy of "writing one to throw away"—or—"write the letter you _want_ to write, then throw that away and write the letter you will actually send." Thank you for indulging my juvenile expressions of frustration.

Updates to follow in the next release … . Thank you again for taking the time to bless your fans with your guidance! 🙇‍♂️❤️
